network/lwip2transport/udp.go

```diff
@@ -15,103 +15,110 @@
+	}
+	_, err := reqSender.WriteTo(data, destAddr)
+	return err
+}
+
+// newSession creates a new UDP session related to conn.
+func (h *udpHandler) newSession(conn lwip.UDPConn) error {
+	laddr := conn.LocalAddr().String()
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if _, ok := h.senders[laddr]; ok {
+		return fmt.Errorf("session already exists for local address %v", laddr)
+	}
+
+	respWriter := &udpConnResponseWriter{conn, h}
+
+	// TODO: we can move this out of h.mu.Lock() to increase performance
+	//       but that requires additional handling of early arrived packets
+	reqSender, err := h.handler.NewSession(respWriter)
 	if err != nil {
+		respWriter.Close()
 		return err
 	}
-	h.Lock()
-	h.conns[tunConn.LocalAddr().String()] = proxyConn
-	h.Unlock()
-	go h.relayPacketsFromProxy(tunConn, proxyConn)
+
+	h.senders[laddr] = reqSender
 	return nil
 }
 
-// relayPacketsFromProxy relays packets from the proxy to the TUN device.
-func (h *udpHandler) relayPacketsFromProxy(tunConn lwip.UDPConn, proxyConn net.PacketConn) {
-	buf := lwip.NewBytes(lwip.BufSize)
-	defer func() {
-		h.close(tunConn)
-		lwip.FreeBytes(buf)
-	}()
-	for {
-		proxyConn.SetDeadline(time.Now().Add(h.timeout))
-		n, sourceAddr, err := proxyConn.ReadFrom(buf)
-		if err != nil {
-			return
-		}
-		// No resolution will take place, the address sent by the proxy is a resolved IP.
-		sourceUDPAddr, err := net.ResolveUDPAddr("udp", sourceAddr.String())
-		if err != nil {
-			return
-		}
-		_, err = tunConn.WriteFrom(buf[:n], sourceUDPAddr)
-		if err != nil {
-			return
-		}
+// closeSession cleans up resources related to conn.
+func (h *udpHandler) closeSession(conn lwip.UDPConn) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	laddr := conn.LocalAddr().String()
+	if reqSender, ok := h.senders[laddr]; ok {
+		reqSender.Close()
+		delete(h.senders, laddr)
 	}
 }
 
-// ReceiveTo relays packets from the TUN device to the proxy. It's called by tun2socks.
-func (h *udpHandler) ReceiveTo(tunConn lwip.UDPConn, data []byte, destAddr *net.UDPAddr) error {
-	h.Lock()
-	proxyConn, ok := h.conns[tunConn.LocalAddr().String()]
-	h.Unlock()
-	if !ok {
-		return fmt.Errorf("connection %v->%v does not exist", tunConn.LocalAddr(), destAddr)
-	}
-	proxyConn.SetDeadline(time.Now().Add(h.timeout))
-	_, err := proxyConn.WriteTo(data, destAddr)
-	return err
+// The PacketResponseWriter that will write responses to the lwip network stack.
+type udpConnResponseWriter struct {
+	conn lwip.UDPConn
+	h    *udpHandler
 }
 
-func (h *udpHandler) close(tunConn lwip.UDPConn) {
-	laddr := tunConn.LocalAddr().String()
-	tunConn.Close()
-	h.Lock()
-	defer h.Unlock()
-	if proxyConn, ok := h.conns[laddr]; ok {
-		proxyConn.Close()
-		delete(h.conns, laddr)
+// Write relays packets from the proxy to the lwIP TUN device. It's called by transport.PacketSession.
+func (r *udpConnResponseWriter) WriteFrom(p []byte, source net.Addr) (int, error) {
+	srcAddr, err := net.ResolveUDPAddr("udp", source.String())
```

Comment:
Why is this resolution needed? 
`r.WriteFrom` takes a `net.Addr`
Please add a comment.

Reply:
It should be `r.conn.WriteFrom`, which takes a `*net.UDPAddr`. will fix the code in next commit.

Reply:
Try `transport.MakeNetAddr`.

Reply:
`transport.MakeNetAddr` returns a `net.Addr`, but we need a `*net.UDPAddr` parsed from `net.Addr`. Then we still need to convert it to `*net.UDPAddr` by using type assertion, it doesn't simplify the code here, so I will just keep it as it is now.